### PR TITLE
CPDP-2583

### PR DESCRIPTION
**Pull request by sam-miele**

### CODE REVIEW

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html [hunk 1]
```diff
@@ -20,44 +20,87 @@
...
+            case Address.NonUkAddress(_, _, _, _, _, _) => ""
+        }
+        case None => ""
+    }
+  }
+
+@line2Display = @{
+    previousReturn match {
+        case Some(value) => value.propertyAddress match {
+            case Address.UkAddress(_, _, _, Some(town), postcode) => s"$town, ${postcode.value}"
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html:59]: do we not want to show the full address?

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html [hunk 2]
```diff
@@ -20,44 +20,87 @@
...
+    <p>@Html(messages(s"$key.p2"))</p>
+
+    <hr/>
+
+    <ul class="return-list">
+        <li>
+            <div class="grid-row">
+                <div class="column-three-quarters submission-summary">
+                    <h3>
+                        <span class="caption">Existing return details</span>
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html:88]: this should come from the messages file

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html [hunk 3]
```diff
@@ -20,44 +20,87 @@
...
+                        @line1Display
+                        <span class="sub-heading">@line2Display</span>
+                    </h3>
+                </div>
+                </div>
+            <a id="viewSentReturn-@submissionId"
+            class="button"
+            role="button"
+            href=@controllers.accounts.homepage.routes.HomePageController.viewSentReturn(submissionId)>View and change return</a>
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html:97]: this should come from the messages file as well

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html [hunk 4]
```diff
@@ -20,44 +20,87 @@
...
+@import uk.gov.hmrc.cgtpropertydisposalsfrontend.controllers
 
 @this(
         mainTemplate: uk.gov.hmrc.cgtpropertydisposalsfrontend.views.html.main_template,
         pageHeading: uk.gov.hmrc.cgtpropertydisposalsfrontend.views.html.components.page_heading,
         backLinkComponent: uk.gov.hmrc.cgtpropertydisposalsfrontend.views.html.components.back_link
 )
 
-@(isATrust: Boolean, backLink: Call)(implicit request: RequestWithSessionData[_], messages: Messages, appConfig: ViewConfig)
+@(previousReturn: Option[ReturnSummary], backLink: Call)(implicit request: RequestWithSessionData[_], messages: Messages, appConfig: ViewConfig)
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html:34]: couldn't `previousReturn` be non-optional. In theory, you should always be able to have a reference to the `ReturnSummary` where the same completion date exists

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/triage/CommonTriageQuestionsController.scala [hunk 5]
```diff
@@ -419,12 +420,59 @@ class CommonTriageQuestionsController @Inject() (
...
+            .fold(
+              _.fold(
+                _.completionDate,
+                e => Some(e.completionDate)
+              ),
+              _.fold(
+                _.completionDate,
+                e => Some(e.completionDate)
+              )
+            )
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/triage/CommonTriageQuestionsController.scala:446]: if `invalidCompletionDate` is `None` you can redirect to either the single or multiple disposal triage cya endpoint depending on what type of triage answers you have - if it's `None` then it means you shouldn't be on this page cos it means no completion date has been supplied to us.

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/triage/CommonTriageQuestionsController.scala [hunk 6]
```diff
@@ -419,12 +420,59 @@ class CommonTriageQuestionsController @Inject() (
...
         )
+        val matching              = state.fold(
+          _.previousSentReturns
+            .map(_.summaries)
+            .getOrElse(List.empty)
+            .filter(e => invalidCompletionDate.contains(CompletionDate(e.completionDate)))
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/triage/CommonTriageQuestionsController.scala:452]: you can replace `.filter(..).headOption` with `.find(...)`

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date_ammend.scala.html [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+@import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.TimeUtils
+@import java.time.LocalDate
+
+@this(
+        mainTemplate: uk.gov.hmrc.cgtpropertydisposalsfrontend.views.html.main_template,
+        pageHeading: uk.gov.hmrc.cgtpropertydisposalsfrontend.views.html.components.page_heading,
+        backLinkComponent: uk.gov.hmrc.cgtpropertydisposalsfrontend.views.html.components.back_link
+)
+
+@(invalidCompletionDate: Option[CompletionDate], isATrust: Boolean, backLink: Call)(implicit request: RequestWithSessionData[_], messages: Messages, appConfig: ViewConfig)
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date_ammend.scala.html:34]: `invalidCompletionDate` can be non-optional here as well

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html (hunk 1 above)
Comment (sam-miele) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html:59]: Noo. The mockups looked like.. 
image

Comment (AndyHWChung) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html:59]: that might just be the way the data was set up with some of the address lines blank. It'd be good to double check if all lines of address should be displayed.

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/triage/CommonTriageQuestionsController.scala [hunk 8]
```diff
@@ -419,15 +420,86 @@ class CommonTriageQuestionsController @Inject() (
...
+            _.previousSentReturns
+              .map(_.summaries)
+              .getOrElse(List.empty)
+              .find(e => CompletionDate(e.completionDate) === value),
+            _.previousSentReturns
+              .map(_.summaries)
+              .getOrElse(List.empty)
+              .find(e => CompletionDate(e.completionDate) === value)
+          ) match {
+            case Some(matchingPreviousReturn) if state.fold(_ => true, r => !r.isAmendReturn) =>
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/triage/CommonTriageQuestionsController.scala:434]: you could use `state.forall(..)` here to make it shorter

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/triage/CommonTriageQuestionsController.scala [hunk 9]
```diff
@@ -419,15 +420,86 @@ class CommonTriageQuestionsController @Inject() (
...
+              Ok(
+                previousReturnExistsWithSameCompletionDatePage(
+                  matchingPreviousReturn,
+                  backLink
+                )
+              )
+            case _ if state.fold(_ => false, _.isAmendReturn)                                 =>
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/triage/CommonTriageQuestionsController.scala:441]: and you could use `state.exists(..)` here as well to make it shorter

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/triage/CommonTriageQuestionsController.scala [hunk 10]
```diff
@@ -419,15 +420,86 @@ class CommonTriageQuestionsController @Inject() (
...
+          }
+        }
       }
     }
 
+  private def withCompletionDate(state: Either[StartingNewDraftReturn, FillingOutReturn])(
+    f: CompletionDate => Future[Result]
+  ): Future[Result] =
+    state match {
+      case Left(value)  =>
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/triage/CommonTriageQuestionsController.scala:459]: having `value` everywhere here is a bit confusing - it'd be better to have more descriptive names

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/triage/CommonTriageQuestionsController.scala [hunk 11]
```diff
@@ -419,15 +420,86 @@ class CommonTriageQuestionsController @Inject() (
...
+            value match {
+              case s: IncompleteSingleDisposalTriageAnswers                           =>
+                s.completionDate match {
+                  case Some(value) => f(value)
+                  case None        => Redirect(routes.SingleDisposalsTriageController.whenWasCompletionDate())
+                }
+              case s: SingleDisposalTriageAnswers.CompleteSingleDisposalTriageAnswers => f(s.completionDate)
+            }
+        }
+    }
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/triage/CommonTriageQuestionsController.scala:501]: there should be some way to factor out the common code in this function by perhaps having  local `def` inside it

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html [hunk 12]
```diff
@@ -20,44 +20,69 @@
...
+        case Address.NonUkAddress(line1, _, _, _, _, _) => line1
+    }
 }
 
-
-@title = @{messages(s"$key.title")}
+@line2Display = @{
+    previousReturn.propertyAddress match {
+        case Address.UkAddress(_, _, _, Some(town), postcode) => s"$town, ${postcode.value}"
+        case _ => ""
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html:48]: did you confirm how uk and non-uk addresses were supposed to be displayed? Are all lines supposed to be there or not?

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html [hunk 13]
```diff
@@ -39,12 +39,14 @@
     previousReturn.propertyAddress match {
         case Address.UkAddress(line1, _, _, _, _) => line1
         case Address.NonUkAddress(line1, _, _, _, _, _) => line1
+        case _ => ""
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html:42]: there is no other case - an address is either a uk address or a non-uk address

File: app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html [hunk 14]
```diff
@@ -39,12 +39,14 @@
...
     }
 }
 
 @line2Display = @{
     previousReturn.propertyAddress match {
         case Address.UkAddress(_, _, _, Some(town), postcode) => s"$town, ${postcode.value}"
+        case Address.NonUkAddress(_, Some(area), Some(townOrCity), _, _, _) => s"$area, $townOrCity"
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html:49]: there is no other case below this line either

Comment (sam-miele) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html:49]: There's a case where it's a NonUKAddress and the optional area/townorcity are None

Comment (AndyHWChung) [reply app/uk/gov/hmrc/cgtpropertydisposalsfrontend/views/returns/triage/previous_return_exists_with_same_completion_date.scala.html:49]: oh yeah,,,what happens if the's a uk address with no town then? Do we still want to display the postcode?

**Comment by AndyHWChung** [comment]
LGTM

**Pull request merged**